### PR TITLE
[6.x] Fix memory usage on downloading large files

### PR DESCRIPTION
**Pull request by sebsel**
Fixes issue #31159

### CODE REVIEW

**Comment by voyula** [comment]
Is it really working?

**Comment by voyula** [comment]
I think "while" is will not solve because output buffering is active and output buffering will import all download data in memory before send to response.

**Comment by sebsel** [comment]
I have just tested it again to be sure, but no: this really works for me and solves the problem.

To be fair, I'm testing on localhost with the local driver, but as I understand it other drivers shouldn't really matter.

**Comment by voyula** [comment]
Hmm, because output buffering is the active, i checked, it is active before send to response...

If you will 100GB+ download data, you need 100GB+ RAM.

**Comment by sebsel** [comment]
Yes, as you understand it. Have you ran the code?

The following:
```php
            $stream = $this->readStream($path);
            while(! feof($stream)) {
                echo fread($stream, 2048);
            }
            fclose($stream);

            \Log::info(memory_get_peak_usage());
```

... with a file download of 1,04 GB. Gives me 21 MB in the log:

```
[2020-01-17 18:41:54] local.INFO: 21088296
```

Do you have an example in which it does not work? (I want to learn too here.)

**Comment by voyula** [comment]
@sebsel 
File size: 2147483648 bytes

\Storage::disk('local_public')->download('vs2015.pro_enu_3.iso')->send();

 `
[2020-01-17 21:51:38] local.ERROR: Allowed memory size of 134217728 bytes exhausted (tried to allocate 2147483648 bytes) {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalErrorException(code: 1): Allowed memory size of 134217728 bytes exhausted (tried to allocate 2147483648 bytes) at D:\\Projects\\vapahy\\leenpa\\vendor\\laravel\\framework\\src\\Illuminate\\Filesystem\\FilesystemAdapter.php:165)
[stacktrace]
#0 {main}
"} 
`

**Comment by sebsel** [comment]
Very interesting, thanks!

The way I tested it was through a `File` Nova-field download. I can reproduce that error too on my Mac here, when I run it in a tinker-session: it streams the first part, and then blows up midway.

From the context I get here, however, I think it's Tinker itself that blows up. See the reference to 'php-console-highlighter':

```
   Symfony\Component\Debug\Exception\FatalErrorException  : Allowed memory size of 134217728 bytes exhausted (tried to allocate 109051936 bytes)

  at /Users/sebastiaan/code/orca/vendor/psy/psysh/src/Shell.php:936
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 266240 bytes) in /Users/sebastiaan/code/orca/vendor/jakub-onderka/php-console-highlighter/src/Highlighter.php on line 113
PHP Fatal error:  Unknown: Cannot use output buffering in output buffering display handlers in Unknown on line 0
```

The crash in the console is also in the log, but there is no mention of crashes in the logs when I go through the browser, so I am still convinced my fix works.

**Comment by sebsel** [comment]
@voyula It's not bad to be wrong. I was wrong too, because I thought you did not run the code and you did and you had a good point. Don't delete your comments. Thanks for the research!

**Comment by voyula** [comment]
@sebsel I dont sure. Yes, it is seems working now! Magically! Output buffering does not affect.

File: src/Illuminate/Filesystem/FilesystemAdapter.php [hunk 1]
```diff
@@ -162,7 +162,9 @@ public function response($path, $name = null, array $headers = [], $disposition
 
         $response->setCallback(function () use ($path) {
             $stream = $this->readStream($path);
-            fpassthru($stream);
+            while (! feof($stream)) {
+                echo fread($stream, 2048);
```
Comment (mfn) [inline src/Illuminate/Filesystem/FilesystemAdapter.php]: Do we know why this "dark magic hackery" fixes it 🧙‍♂️ ?

Comment (voyula) [reply src/Illuminate/Filesystem/FilesystemAdapter.php]: @mfn I was researched now, fpassthru will write all contents to output buffer.

> Like

```php
file_put_contents('php://output', "All file data here");
```
In here, **fpassthru** will get all file data for sending to output buffer, so will temporary memory usage till send process all data into output buffer done.

We can optimize this with chunk via fread.
```php
while (! feof($stream)) {
    file_put_contents('php://output', fread($stream, 2048));
}
```

**Comment by voyula** [comment]
**This PR not working, if the download data is non-binary.**
You can test with text format mock download data.
Will conflict with output buffering.
Will in-memory via output buffering.

File: src/Illuminate/Filesystem/FilesystemAdapter.php (hunk 1 above)
Comment (voyula) [reply src/Illuminate/Filesystem/FilesystemAdapter.php]: Seems like output buffering tries chunking data with php free memory size.
if you will pass upper than free memory, output buffering will be failed.
This `while` giving data to output buffer with small sizes for help to output buffering chunks.

Comment (LastDragon-ru) [reply src/Illuminate/Filesystem/FilesystemAdapter.php]: Why not stream_copy_to_stream? Eg https://github.com/symfony/http-foundation/blob/0f977dd6a6ab22789f49cefccdb623734da2ba3c/BinaryFileResponse.php#L297-L303

Comment (voyula) [reply src/Illuminate/Filesystem/FilesystemAdapter.php]: > Why not stream_copy_to_stream? Eg https://github.com/symfony/http-foundation/blob/0f977dd6a6ab22789f49cefccdb623734da2ba3c/BinaryFileResponse.php#L297-L303

Same things actually.

**Comment by taylorotwell** [comment]
Reverting for breaking downloads.

**Comment by LucaColombi** [comment]
> Reverting for breaking downloads.

hello, had the same problem, memory overflow while handling a large set of data, in 
......... /php-console-highlighter/src/Highlighter.php on line 113


using "laravel/framework": "5.8"


so, this fix does not work because it break something else? Or can I apply it? have to apply it manually in sources?

**Comment by driesvints** [comment]
@LucaColombi 5.8 isn't supported anymore, please upgrade.

**Comment by rtrudel** [comment]
Same problem with Laravel 7. No fix yet?

**Comment by driesvints** [comment]
@rtrudel you're free to do another attempt if you like.

**Comment by rik471** [comment]
Going through the same problem, due to the revert it wasn't fixed yet, any update about whether it's going to be fixed in next versions? I'm thinking on overriding the method for now, anyone got some work around?

**Comment by rik471** [comment]
> Going through the same problem, due to the revert it wasn't fixed yet, any update about whether it's going to be fixed in next versions? I'm thinking on overriding the method for now, anyone got some work around?

I'd like to share a workaround I got based on the "fix" @sebsel introduced.
If your using `streamDownload()` ,`stream()` or even `setCallback()` methods for downloading big files, instead of just echoing the file path on the download method callback, just use something like this inside the callback of the methods mentioned:
```
return response()->streamDownload(function () use ($filePath) {
                    $fd = fopen($filePath, 'rb');
                    while(!feof($fd)) {
                        echo fread($fd, 2048);
                    }
                }, $fileName);
```

The above is chunking the stream of the file instead of loading it through php memory, as mentioned before by @sebsel.
I'm just sharing this so people won't be thinking it's a Laravel bug in which we can't do nothing about it. 
Using the above means your isolating the download to chunks only in the method you need it.
Just trying to help.

Got this fix here: https://www.php.net/manual/en/function.fpassthru.php#29162

**Comment by BlackbitDevs** [comment]
> Reverting for breaking downloads.

What exactly did this change break?

**Pull request merged**